### PR TITLE
Add context-aware functions to vault/api

### LINKED ISSUE #14070: Support context propagation in http clients
**Is your feature request related to a problem? Please describe.**
In the case of long requests, it is impossible to terminate them via context cancellation.

**Describe the solution you'd like**
Add methods with context propagation, e.g. 
- `ReadContext(ctx context.Context, path string) (*Secret, error)`
- `ListContext(ctx context.Context, path string) (*Secret, error)`
- `WriteContext(ctx context.Context, path string, data map[string]interface{}) (*Secret, error)`
- etc.

**Describe alternatives you've considered**
No alternatives in the context of Golang best practices.

**Explain any additional use-cases**
It can take a long time for a certificate to be generated in the Vault. It may be necessary to abort the request and terminate the job correctly.

**Additional context**
No change of existing methods/signatures is required. New methods will be added. The only difference is using the context passed to the method as a parameter instead of `context.Background()`.

### PR DESCRIPTION
**Pull request by averche**
### Credits

This pull request is based on a community contribution #14152 by [ilyakaznacheev](https://github.com/ilyakaznacheev) and resolves #14070

### Description

For each API wrapper function in `vault/api`, we are adding a context-aware equivalent. These new functions will have `WithContext` appended to their names.

### Implementation Notes

One of the issues that I discovered in the original PR (#14152) is a potential context leak due to this piece of code inside `RawRequestWithContext`:

```go
	if timeout != 0 {
		// Note: we purposefully do not call cancel manually. The reason is
		// when canceled, the request.Body will EOF when reading due to the way
		// it streams data in. Cancel will still be run when the timeout is
		// hit, so this doesn't really harm anything.
		ctx, _ = context.WithTimeout(ctx, timeout)
	}
```

To prevent context leaks, the timeout logic has been extracted out of `RawRequestWithContext` and into each `...WithContext` wrapper (using `withConfiguredTimeout` helper).  Additionally, since `RawRequestWithContext` is still used outside of `vautl/api`, its implementation has been moved into unexported `rawRequestWithContext`

See [this comment](https://github.com/hashicorp/vault/pull/14152#r810400167) for more details.
 
### Testing

All existing tests that were calling vault/api functions have been switched to call `WithContext` equivalents with `context.Background()` passed in. Since `WithContext` functions invoke the regular ones, this change should increase code coverage without any negative side effects.

### CODE REVIEW

File: api/sys_monitor.go [hunk 1]
```diff
@@ -20,7 +17,7 @@ func (c *Sys) Monitor(ctx context.Context, logLevel string) (chan string, error)
 		r.Params.Add("log_level", logLevel)
 	}
 
-	resp, err := c.c.rawRequestWithContext(ctx, r)
+	resp, err := c.c.RawRequestWithContext(ctx, r)
```
Comment (averche) [inline api/sys_monitor.go:20]: I had to revert the changes to Monitor(ctx, logLevel) function. This function is asynchronous - it starts a go routine which escapes the scope of the function. The go routine waits for the context to be cancelled before closing. Therefore, we cannot cancel the context within the function since it would immediately cause the go routine to exit.

Thanks unit tests for catching the bug!

File: api/sys_mounts.go [hunk 2]
```diff
@@ -55,27 +65,37 @@ func (c *Sys) Mount(path string, mountInfo *MountInput) error {
...
+	return c.UnmountWithContext(context.Background(), path)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) UnmountWithContext(ctx context.Context, path string) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("DELETE", fmt.Sprintf("/v1/sys/mounts/%s", path))
```
Comment (VinnyHC) [inline api/sys_mounts.go:75]:
```suggestion
	r := c.c.NewRequest(http.MethodDelete, fmt.Sprintf("/v1/sys/mounts/%s", path))
```

File: api/sys_stepdown.go [hunk 3]
```diff
@@ -3,11 +3,16 @@ package api
...
+	return c.StepDownWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) StepDownWithContext(ctx context.Context) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("PUT", "/v1/sys/step-down")
```
Comment (VinnyHC) [inline api/sys_stepdown.go:13]:
```suggestion
	r := c.c.NewRequest(http.MethodPut, "/v1/sys/step-down")
```

File: api/sys_seal.go [hunk 4]
```diff
@@ -3,57 +3,82 @@ package api
...
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) SealWithContext(ctx context.Context) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
-	if err == nil {
-		defer resp.Body.Close()
+
+	r := c.c.NewRequest("PUT", "/v1/sys/seal")
```
Comment (VinnyHC) [inline api/sys_seal.go:22]:
```suggestion
	r := c.c.NewRequest(http.MethodPut, "/v1/sys/seal")
```

File: api/sys_raft.go [hunk 5]
```diff
@@ -235,13 +235,19 @@ func (c *Sys) RaftSnapshotRestoreWithContext(ctx context.Context, snapReader io.
...
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+// RaftAutopilotStateWithContext returns the state of the raft cluster as seen by autopilot.
+func (c *Sys) RaftAutopilotStateWithContext(ctx context.Context) (*AutopilotState, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/autopilot/state")
```
Comment (VinnyHC) [inline api/sys_raft.go:248]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/storage/raft/autopilot/state")
```

File: api/sys_raft.go [hunk 6]
```diff
@@ -269,13 +275,19 @@ func (c *Sys) RaftAutopilotState() (*AutopilotState, error) {
...
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+// RaftAutopilotConfigurationWithContext fetches the autopilot config.
+func (c *Sys) RaftAutopilotConfigurationWithContext(ctx context.Context) (*AutopilotConfig, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/autopilot/configuration")
```
Comment (VinnyHC) [inline api/sys_raft.go:288]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/storage/raft/autopilot/configuration")
```

File: api/sys_policy.go [hunk 7]
```diff
@@ -41,11 +46,16 @@ func (c *Sys) ListPolicies() ([]string, error) {
...
+	return c.GetPolicyWithContext(context.Background(), name)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) GetPolicyWithContext(ctx context.Context, name string) (string, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", fmt.Sprintf("/v1/sys/policies/acl/%s", name))
```
Comment (VinnyHC) [inline api/sys_policy.go:56]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, fmt.Sprintf("/v1/sys/policies/acl/%s", name))
```

File: api/sys_policy.go [hunk 8]
```diff
@@ -93,11 +108,16 @@ func (c *Sys) PutPolicy(name, rules string) error {
...
+	return c.DeletePolicyWithContext(context.Background(), name)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) DeletePolicyWithContext(ctx context.Context, name string) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("DELETE", fmt.Sprintf("/v1/sys/policies/acl/%s", name))
```
Comment (VinnyHC) [inline api/sys_policy.go:118]:
```suggestion
	r := c.c.NewRequest(http.MethodDelete, fmt.Sprintf("/v1/sys/policies/acl/%s", name))
```

File: api/sys_mounts.go [hunk 9]
```diff
@@ -125,13 +151,19 @@ func (c *Sys) StartRemount(from, to string) (*MountMigrationOutput, error) {
...
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+// RemountStatusWithContext checks the status of a mount migration operation with the provided ID
+func (c *Sys) RemountStatusWithContext(ctx context.Context, migrationID string) (*MountMigrationStatusOutput, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", fmt.Sprintf("/v1/sys/remount/status/%s", migrationID))
```
Comment (VinnyHC) [inline api/sys_mounts.go:164]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, fmt.Sprintf("/v1/sys/remount/status/%s", migrationID))
```

File: api/sys_leases.go [hunk 10]
```diff
@@ -57,40 +74,55 @@ func (c *Sys) Revoke(id string) error {
...
+	return c.RevokeForceWithContext(context.Background(), id)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) RevokeForceWithContext(ctx context.Context, id string) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("PUT", "/v1/sys/leases/revoke-force/"+id)
```
Comment (VinnyHC) [inline api/sys_leases.go:109]:
```suggestion
	r := c.c.NewRequest(http.MethodPut, "/v1/sys/leases/revoke-force/"+id)
```

File: api/sys_leases.go [hunk 11]
```diff
@@ -57,40 +74,55 @@ func (c *Sys) Revoke(id string) error {
...
+	return c.RevokePrefixWithContext(context.Background(), id)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) RevokePrefixWithContext(ctx context.Context, id string) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("PUT", "/v1/sys/leases/revoke-prefix/"+id)
```
Comment (VinnyHC) [inline api/sys_leases.go:92]:
```suggestion
	r := c.c.NewRequest(http.MethodPut, "/v1/sys/leases/revoke-prefix/"+id)
```

File: api/sys_leader.go [hunk 12]
```diff
@@ -6,11 +6,16 @@ import (
...
+	return c.LeaderWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) LeaderWithContext(ctx context.Context) (*LeaderResponse, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/leader")
```
Comment (VinnyHC) [inline api/sys_leader.go:16]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/leader")
```

File: api/sys_init.go [hunk 13]
```diff
@@ -3,11 +3,16 @@ package api
...
+	return c.InitStatusWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) InitStatusWithContext(ctx context.Context) (bool, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/init")
```
Comment (VinnyHC) [inline api/sys_init.go:13]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/init")
```

File: api/sys_generate_root.go [hunk 14]
```diff
@@ -3,23 +3,36 @@ package api
...
+func (c *Sys) GenerateRecoveryOperationTokenStatusWithContext(ctx context.Context) (*GenerateRootStatusResponse, error) {
+	return c.generateRootStatusCommonWithContext(ctx, "/v1/sys/generate-recovery-token/attempt")
+}
+
+func (c *Sys) generateRootStatusCommonWithContext(ctx context.Context, path string) (*GenerateRootStatusResponse, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", path)
```
Comment (VinnyHC) [inline api/sys_generate_root.go:33]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, path)
```

File: api/sys_config_cors.go [hunk 15]
```diff
@@ -36,26 +41,36 @@ func (c *Sys) CORSStatus() (*CORSResponse, error) {
...
+	return c.DisableCORSWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) DisableCORSWithContext(ctx context.Context) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("DELETE", "/v1/sys/config/cors")
```
Comment (VinnyHC) [inline api/sys_config_cors.go:71]:
```suggestion
	r := c.c.NewRequest(http.MethodDelete, "/v1/sys/config/cors")
```

File: api/sys_config_cors.go [hunk 16]
```diff
@@ -8,11 +8,16 @@ import (
...
+	return c.CORSStatusWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) CORSStatusWithContext(ctx context.Context) (*CORSResponse, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/config/cors")
```
Comment (VinnyHC) [inline api/sys_config_cors.go:18]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/config/cors")
```

File: api/sys_auth.go [hunk 17]
```diff
@@ -9,11 +9,16 @@ import (
...
+	return c.ListAuthWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) ListAuthWithContext(ctx context.Context) (map[string]*AuthMount, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/auth")
```
Comment (VinnyHC) [inline api/sys_auth.go:19]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/auth")
```

File: api/sys_audit.go [hunk 18]
```diff
@@ -102,11 +117,16 @@ func (c *Sys) EnableAuditWithOptions(path string, options *EnableAuditOptions) e
...
+	return c.DisableAuditWithContext(context.Background(), path)
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) DisableAuditWithContext(ctx context.Context, path string) error {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("DELETE", fmt.Sprintf("/v1/sys/audit/%s", path))
```
Comment (VinnyHC) [inline api/sys_audit.go:127]:
```suggestion
	r := c.c.NewRequest(http.MethodDelete, fmt.Sprintf("/v1/sys/audit/%s", path))
```

File: api/sys_audit.go [hunk 19]
```diff
@@ -47,11 +52,16 @@ func (c *Sys) AuditHash(path string, input string) (string, error) {
...
+	return c.ListAuditWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *Sys) ListAuditWithContext(ctx context.Context) (map[string]*Audit, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/sys/audit")
```
Comment (VinnyHC) [inline api/sys_audit.go:62]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/sys/audit")
```

File: api/auth_token.go [hunk 20]
```diff
@@ -104,11 +129,16 @@ func (c *TokenAuth) LookupAccessor(accessor string) (*Secret, error) {
...
+	return c.LookupSelfWithContext(context.Background())
+}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+func (c *TokenAuth) LookupSelfWithContext(ctx context.Context) (*Secret, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
 	defer cancelFunc()
-	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	r := c.c.NewRequest("GET", "/v1/auth/token/lookup-self")
```
Comment (VinnyHC) [inline api/auth_token.go:139]:
```suggestion
	r := c.c.NewRequest(http.MethodGet, "/v1/auth/token/lookup-self")
```

**Review by VinnyHC — COMMENTED** [review]
This looks really good!
Tiny nit: I added suggestions to move any new code over to using the `http` packages constants instead of strings but this might be best as a different PR that does it across the entire api library.

**Review by VinnyHC — APPROVED** [review]
Nice! This looks really good to me. Nice work moving the tests over.

File: api/sys_init.go [hunk 21]
```diff
@@ -19,14 +27,19 @@ func (c *Sys) InitStatus() (bool, error) {
...
 
 func (c *Sys) Init(opts *InitRequest) (*InitResponse, error) {
+	return c.InitWithContext(context.Background(), opts)
+}
+
+func (c *Sys) InitWithContext(ctx context.Context, opts *InitRequest) (*InitResponse, error) {
+	ctx, cancelFunc := c.c.withConfiguredTimeout(ctx)
+	defer cancelFunc()
+
 	r := c.c.NewRequest("PUT", "/v1/sys/init")
```
Comment (digivava) [inline api/sys_init.go:37]: I notice in several places you're replacing "GET" with http.MethodGet, but I don't see it for the PUTs here

File: api/sys_capabilities.go [hunk 22]
```diff
@@ -12,7 +12,21 @@ func (c *Sys) CapabilitiesSelf(path string) ([]string, error) {
 	return c.Capabilities(c.c.Token(), path)
```
Comment (digivava) [inline api/sys_capabilities.go:12]: This one isn't calling the WithContext version of itself like the others, but that's fine because it calls Capabilities which creates a context.Background(), right? Meaning CapabilitiesSelfWithContext is only used by those users who invoke it directly.

File: api/sys_init.go (hunk 21 above)
Comment (averche) [reply api/sys_init.go:37]: I wanted to do another PR to go over all of the vault/api methods and replace the strings with constants. I merged the ones from Vinny's suggestions only for the already modified lines.

File: api/sys_capabilities.go (hunk 22 above)
Comment (averche) [reply api/sys_capabilities.go:12]: Great catch! 👍 This is technically correct but it's good to standardize the code, I'll fix it.

**Comment by averche** [comment]
> Whew, good work! LGTM, aside from some inconsistency in when the http.MethodGet is used vs the hard-coded "GET".

Good point, I think I'll just revert all of the `http.MethodGet` / `http.MethodPut` changes and create another PR for that after this one. It's not a good to mix two things in one PR (especially a huge one like this).

> I am starting to second-guess whether adding such a large amount of duplicate methods is a better choice than just adding ctx to all the existing ones and bumping to a new major version (where breaking changes would presumably be expected)... but it's a little late for that now so we can rethink that in the new version of the client I suppose!

This is meant as a stop-gap solution until we have a new shiny version. In the new version, we'll most likely default to context-only functions.

File: api/sys_init.go (hunk 21 above)
Comment (averche) [reply api/sys_init.go:34]: On a second thought, I decided to revert the http.MethodGet/Put changes and will do them in a separate PR.
